tensorflow_quantum/core/ops/math_ops/tfq_inner_product_adj_grad.cc

```diff
@@ -0,0 +1,478 @@
+                            scratch2);
+
+            // don't need not-found check since this is done upstream already.
+            const auto it = maps[i].find(gradient_gates[i][l - 1].params[k]);
+            const int loc = it->second.first;
+            // Apply finite differencing for adjoint gradients.
+            // Finite differencing enables applying multiple `gradient_gate`
+            // of a symbol at the same circuit. For analytic methods like
+            // parameter-shift we need to apply a single `gradient_gate`
+            // per a symbol.
+            std::complex<double> result = ss.InnerProduct(scratch2, scratch);
+            (*output_tensor)(i, j, loc) +=
+                std::complex<float>(static_cast<float>(result.real()),
+                                    static_cast<float>(result.imag()));
+          }
+          ApplyGateDagger(sim, cur_gate, scratch);
+        }
+      }
+    }
+  }
+
+  void ComputeSmall(
+      const std::vector<int>& num_qubits, const int max_num_qubits,
+      const std::vector<SymbolMap>& maps,
+      const std::vector<QsimCircuit>& qsim_circuits,
+      const std::vector<QsimFusedCircuit>& fused_circuits,
+      const std::vector<std::vector<std::vector<qsim::GateFused<QsimGate>>>>&
+          partial_fused_circuits,
+      const std::vector<std::vector<tfq::GradientOfGate>>& gradient_gates,
+      const std::vector<std::vector<QsimFusedCircuit>>& other_fused_circuits,
+      tensorflow::OpKernelContext* context,
+      tensorflow::TTypes<std::complex<float>, 3>::Tensor* output_tensor) {
+    const auto tfq_for = qsim::SequentialFor(1);
+    using Simulator = qsim::Simulator<const qsim::SequentialFor&>;
+    using StateSpace = Simulator::StateSpace;
+
+    const int output_dim_internal_size = output_tensor->dimension(1);
+
+    auto DoWork = [&](int start, int end) {
+      int old_batch_index = -2;
+      int cur_batch_index = -1;
+      int largest_nq = 1;
+      int cur_internal_index;
+
+      Simulator sim = Simulator(tfq_for);
+      StateSpace ss = StateSpace(tfq_for);
+      auto sv = ss.Create(largest_nq);
+      auto sv_adj = ss.Create(largest_nq);
+      auto scratch = ss.Create(largest_nq);
+      auto scratch2 = ss.Create(largest_nq);
+      for (int i = start; i < end; i++) {
+        cur_batch_index = i / output_dim_internal_size;
+        cur_internal_index = i % output_dim_internal_size;
+
+        const int nq = num_qubits[cur_batch_index];
+
+        if (cur_batch_index != old_batch_index) {
+          // We've run into a new state vector we must compute.
+          // Only compute a new state vector when we have to.
+          if (nq > largest_nq) {
+            largest_nq = nq;
+            sv = ss.Create(largest_nq);
+            sv_adj = ss.Create(largest_nq);
+            scratch = ss.Create(largest_nq);
+            scratch2 = ss.Create(largest_nq);
+          }
+          // no need to update scratch_state since ComputeExpectation
```

Comment:
Is ComputeExpectation still here ?

Reply:
Oh, wrong comments. thank you for the finding. let me remove it.